[libcxxabi] Use __LDBL_MANT_DIG__ for configuring demangling of long doubles

mstorsjo opened the pull request:
This avoids needing to hardcode the mapping between architectures and their sizes of long doubles.

This fixes a case in test_demangle.pass.cpp, that previously failed like this (XFAILed):

    .---command stdout------------
    | Testing 29859 symbols.
    | _ZN5test01hIfEEvRAcvjplstT_Le4001a000000000000000E_c should be invalid but is not
    | Got: 0, void test0::h<float>(char (&) [(unsigned int)(sizeof (float) + 0x0.07ff98f7ep-1022L)])
    `-----------------------------
    .---command stderr------------
    | Assertion failed: !passed && "demangle did not fail", file libcxxabi/test/test_demangle.pass.cpp, line 30338
    `-----------------------------

This testcase is defined within

    // Is long double fp80?  (Only x87 extended double has 64-bit mantissa)
    #define LDBL_FP80 (__LDBL_MANT_DIG__ == 64)
    ...
    #if !LDBL_FP80
    ...
    #endif

The case failed on x86 architectures with an unusual size for long doubles, as the test expected the demangler to not be able to demangle an 80 bit long double (based on the `__LDBL_MANT_DIG__ == 64` condition in the test). However as the libcxxabi implementation was hardcoded to demangle 80 bit long doubles on x86_64 regardless of the actual size, this test failed (by unexpectedly being able to demangle it).

By configuring libcxxabi's demangling of long doubles to match what the compiler specifies, we no longer hit the expected failures in the test_demangle.pass.cpp test on Android on x86.

This makes libcxxabi require a GCC-compatible compiler that defines nonstandard defines like `__LDBL_MANT_DIG__`, but I presume that's already essentially required anyway.

mstorsjo committed:
[libcxxabi] Use __LDBL_MANT_DIG__ for configuring demangling of long doubles

This avoids needing to hardcode the mapping between architectures
and their sizes of long doubles.

This fixes a case in test_demangle.pass.cpp, that previously failed
like this (XFAILed):

    .---command stdout------------
    | Testing 29859 symbols.
    | _ZN5test01hIfEEvRAcvjplstT_Le4001a000000000000000E_c should be invalid but is not
    | Got: 0, void test0::h<float>(char (&) [(unsigned int)(sizeof (float) + 0x0.07ff98f7ep-1022L)])
    `-----------------------------
    .---command stderr------------
    | Assertion failed: !passed && "demangle did not fail", file libcxxabi/test/test_demangle.pass.cpp, line 30338
    `-----------------------------

This testcase is defined within

    // Is long double fp80?  (Only x87 extended double has 64-bit mantissa)
    #define LDBL_FP80 (__LDBL_MANT_DIG__ == 64)
    ...
    #if !LDBL_FP80
    ...
    #endif

The case failed on x86 architectures with an unusual size for
long doubles, as the test expected the demangler to not be able to
demangle an 80 bit long double (based on the "_LDBL_MANT_DIG__ == 64"
condition in the test). However as the libcxxabi implementation was
hardcoded to demangle 80 bit long doubles on x86_64 regardless of
the actual size, this test failed (by unexpectedly being able to
demangle it).

By configuring libcxxabi's demangling of long doubles to match
what the compiler specifies, we no longer hit the expected failures
in the test_demangle.pass.cpp test on Android on x86.

This makes libcxxabi require a GCC-compatible compiler that defines
nonstandard defines like __LDBL_MANT_DIG__, but I presume that's
already essentially required anyway.

philnik777 commented on libcxxabi/src/demangle/ItaniumDemangle.h:
```diff
@@ -5741,14 +5741,12 @@ struct FloatData<double>
+  static const size_t mangled_size = 32;
+#elif __LDBL_MANT_DIG__ == 53
+  static const size_t mangled_size = 16;
+#else // __LDBL_MANT_DIG__ == 64
```
I think it'd be better to check the mantissa here as well and `#error` if there is a new size. Otherwise we might be silently wrong on some platforms.

mstorsjo replied:
That sounds like a good idea, will fix.

However - I realized that while we don't support building libcxxabi with compilers that aren't GCC compatible, we also do copy this file into `llvm/include/llvm/Demangle/ItaniumDemangle.h`, which is built as part of the host build of LLVM - and that can be built with MSVC as well. (And MSVC doesn't define any of the arch defines that we defined before either.)

MSVC only has `long double == double` for all currently known architectures (and clang-cl matches this), so I can add an ` || defined(_MSC_VER)` for that.

mstorsjo committed:
Actually check the last case, add an #error if the size is unhandled

Also check for _MSC_VER for MSVC, as this file is copied and included
in the main LLVM build as well.

rprichard reviewed (approved):
Looks good to me.

(FWIW, on x86-64 Android, Clang doesn't emit the `e` mangling -- instead, it emits `g` for both `long double` and `__float128`. https://godbolt.org/z/Gvqof41xz)

philnik777 reviewed (approved):
LGTM assuming the CI is happy.

thurstond commented:
This is failing on the ppc64le-linux buildbot (https://lab.llvm.org/buildbot/#/builders/72/builds/10061):

```
In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/libcxxabi/src/cxa_demangle.cpp:17:
/home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/libcxxabi/src/demangle/ItaniumDemangle.h:5753:2: error: Unknown size for __LDBL_MANT_DIG__
 5753 | #error Unknown size for __LDBL_MANT_DIG__
...
```

mstorsjo commented:
> This is failing on the ppc64le-linux buildbot (https://lab.llvm.org/buildbot/#/builders/72/builds/10061):
> 
> ```
> In file included from /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/libcxxabi/src/cxa_demangle.cpp:17:
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-sanitizer/sanitizer-ppc64le/build/llvm-project/libcxxabi/src/demangle/ItaniumDemangle.h:5753:2: error: Unknown size for __LDBL_MANT_DIG__
>  5753 | #error Unknown size for __LDBL_MANT_DIG__
> ...
> ```

Sorry about that. I see that powerpc has got `__LDBL_MANT_DIG__` 106, contrary to other 128 bit long doubles that have `__LDBL_MANT_DIG__` equal to 113. I guess the correct case for that would be `mangled_size = 32;` here? The old preexisting ifdefs didn't match for powerpc at all (so it got the default fallback case of `mangled_size = 20` which probably isn't right).

Anyway, we need to fix this. Do you suggest a revert, or fix forward like this?
```diff
diff --git a/libcxxabi/src/demangle/ItaniumDemangle.h b/libcxxabi/src/demangle/ItaniumDemangle.h
index b9d17060c04f..eca9ddad66f9 100644
--- a/libcxxabi/src/demangle/ItaniumDemangle.h
+++ b/libcxxabi/src/demangle/ItaniumDemangle.h
@@ -5741,7 +5741,7 @@ struct FloatData<double>
 template <>
 struct FloatData<long double>
 {
-#if __LDBL_MANT_DIG__ == 113
+#if __LDBL_MANT_DIG__ == 113 || __LDBL_MANT_DIG__ == 106
   static const size_t mangled_size = 32;
 #elif __LDBL_MANT_DIG__ == 53 || defined(_MSC_VER)
   // MSVC doesn't define __LDBL_MANT_DIG__, but it has long double equal to
```

thurstond commented:
Thanks for landing the fix-forward so quickly! Sorry for not being able to reply earlier - I had gone to bed (timezone difference).

Michael137 commented:
FYI, can you run ` ./libcxxabi/src/demangle/cp-to-llvm.sh` so the changes are reflected in the LLVM copy of the demangler?

Michael137 commented:
> FYI, can you run ` ./libcxxabi/src/demangle/cp-to-llvm.sh` so the changes are reflected in the LLVM copy of the demangler?

https://github.com/llvm/llvm-project/pull/135968
